src/validation.cpp

```diff
@@ -3902,18 +3904,21 @@ bool ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& blo
         bool ret = CheckBlock(*block, state, GetConsensus());
         if (ret) {
             // Store to disk
-            ret = ActiveChainstate().AcceptBlock(block, state, &pindex, force_processing, nullptr, new_block, min_pow_checked);
+            ret = chainstate->AcceptBlock(
+                block, state, &pindex, force_processing, nullptr, new_block, min_pow_checked);
         }
         if (!ret) {
             GetMainSignals().BlockChecked(*block, state);
             return error("%s: AcceptBlock FAILED (%s)", __func__, state.ToString());
         }
     }
 
-    NotifyHeaderTip(ActiveChainstate());
+    if (chainstate == &this->ActiveChainstate()) {
+        NotifyHeaderTip(*chainstate);
+    }
 
     BlockValidationState state; // Only used to report errors, not invalidity - ignore it
-    if (!ActiveChainstate().ActivateBestChain(state, block)) {
+    if (!chainstate->ActivateBestChain(state, block)) {
```

Comment:
In commit "validation: introduce ChainstateManager::GetChainstateForNewBlock" (650677dd66cede06e242975f51c0e5e77b99079f)

If I revert this line, all tests seem to still pass. I understand why we can't have functional tests for this code before there is a snapshot loading RPC but it seems like it shouldn't be that hard to come up with a unit test that would exercise this logic. I don't think it's good to be in a state where validation code doesn't have test coverage, even if it's a temporary state.